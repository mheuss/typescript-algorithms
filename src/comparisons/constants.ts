export enum Comparisons {
  LessThan = -1,
  Equal,
  GreaterThan,
}

export type ComparisonFunction<T> = (a: T, b: T) => Comparisons;
