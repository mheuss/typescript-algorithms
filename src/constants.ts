export enum ErrorCodes {
  COMPARISON_FUNCTION_REQUIRED = 1,
  LIST_IS_BROKEN_INTERNALLY,
  OPERATION_BEYOND_BOUNDS,
  FAILED_TO_SET,
  COMPARISON_ERROR,
  INVALID_TREE_DATA_FILE,
}
