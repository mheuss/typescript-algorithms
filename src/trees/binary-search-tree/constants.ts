export enum Sides {
  Left,
  Right,
}
