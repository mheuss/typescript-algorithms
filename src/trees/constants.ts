export interface ITreeObject<T> {
  children: Array<ITreeObject<T>>;
  value: T;
}

export enum Traversal {
  Breadth_First,
  DepthFirstInOrder,
  DepthFirstPreOrder,
  DepthFirstPostOrder,
}
