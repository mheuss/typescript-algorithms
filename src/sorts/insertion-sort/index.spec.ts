import { ErrorCodes } from "../../constants";
import { InsertionSort } from "./index";

describe("Insertion sort", () => {
  it("Should sort an array of numbers quite readily", () => {
    const arr: number[] = [5, 2, 8, 6, 1, 3, 7, 9, 4, 10];
    const results = InsertionSort<number>(arr);
    expect(results.length).toEqual(arr.length);
    expect(results).toMatchSnapshot();
  });

  it("Should tackle an array of strings", () => {
    const arr: string[] = [
      "pineapple",
      "artichoke",
      "celery",
      "eggplant",
      "banana",
      "Kiwi",
    ];
    const results = InsertionSort<string>(arr);
    expect(results.length).toEqual(arr.length);
    expect(results).toMatchSnapshot();
  });

  it("Should be able to support an array of objects using a custom sort", () => {
    interface INames {
      lastName: string;
      firstName: string;
    }

    const arr = [
      { lastName: "Heuss", firstName: "Mike" },
      { lastName: "Smith", firstName: "John" },
      { lastName: "Arnold", firstName: "Peggy" },
      { lastName: "Saben", firstName: "Matt" },
      { lastName: "LaCroix", firstName: "Mark" },
    ];
    const results = InsertionSort<INames>(arr, (a: INames, b: INames) => {
      return a.lastName.localeCompare(b.lastName);
    });
    expect(results.length).toEqual(arr.length);
    expect(results).toMatchSnapshot();
  });

  it("Should return an empty array if given an empty array", () => {
    const results = InsertionSort<number>([]);
    expect(results).toEqual([]);
  });

  it("Should throw if array is of type it cannot sort, and no sort function was passed in", () => {
    const arr = [
      { lastName: "Heuss", firstName: "Mike" },
      { lastName: "Smith", firstName: "John" },
      { lastName: "Arnold", firstName: "Peggy" },
      { lastName: "Saben", firstName: "Matt" },
      { lastName: "LaCroix", firstName: "Mark" },
    ];
    try {
      InsertionSort(arr);
      fail("Should have thrown");
    } catch (e) {
      // @ts-expect-error - Not worried about getting the correct type here.
      expect(e.code).toEqual(ErrorCodes.COMPARISON_FUNCTION_REQUIRED);
    }
  });
});
