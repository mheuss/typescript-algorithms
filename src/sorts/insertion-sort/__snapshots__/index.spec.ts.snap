// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Insertion sort Should be able to support an array of objects using a custom sort 1`] = `
Array [
  Object {
    "firstName": "Peggy",
    "lastName": "Arnold",
  },
  Object {
    "firstName": "Mike",
    "lastName": "Heuss",
  },
  Object {
    "firstName": "Mark",
    "lastName": "LaCroix",
  },
  Object {
    "firstName": "Matt",
    "lastName": "Saben",
  },
  Object {
    "firstName": "John",
    "lastName": "Smith",
  },
]
`;

exports[`Insertion sort Should sort an array of numbers quite readily 1`] = `
Array [
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
]
`;

exports[`Insertion sort Should tackle an array of strings 1`] = `
Array [
  "artichoke",
  "banana",
  "celery",
  "eggplant",
  "Kiwi",
  "pineapple",
]
`;
